Add host field to wakes locations

chrismaddocks opened the pull request:
https://trello.com/c/noYZ6n6j/2020-add-hostname-support-to-wakes-locations

benhutton commented on app/models/concerns/wakes/redirectors.rb:
```diff
@@ -16,7 +16,7 @@ def create_redirect_graph
         legacy_locations.reload
         legacy_paths_in_redis = legacy_locations.map do |legacy_location|
-          Wakes::REDIS.set(legacy_location.path, canonical_location.path)
+          Wakes::REDIS.set(legacy_location.path, canonical_location.path) unless legacy_location.host.present?
```
@chrismaddocks i believe that here you need to worry about a host being set on the `canonical_location`. If it's set, we want to redirect to the url, not the path, right?

chrismaddocks replied:
@benhutton Correct—I've now addressed this in the latest commits.

benhutton commented on app/models/wakes/location.rb:
```diff
@@ -3,7 +3,8 @@ class Wakes::Location < ActiveRecord::Base
 
-  validates :path, :format => { :with => %r{\A\/} }, :uniqueness => true
+  validates :path, :format => { :with => %r{\A\/} }
+  validates_uniqueness_of :path, :scope => :host
```
I think you can keep the old format, and use `validates :uniqueness => { :scope => :host }`

benhutton commented on app/models/wakes/location.rb:
```diff
@@ -13,7 +14,8 @@ def label
 
-  def url(protocol: 'http', host: ENV['DEFAULT_HOST'])
+  def url(protocol: 'http', host: nil)
+    host ||= self.host || ENV['DEFAULT_HOST']
```
this is a rather oblique line of code.

Your goal, if I understand it, is to do something like `host || self.host || ENV['DEFAULT_HOST']`, right?

@gaganawhad do you know whether we really need to pass in the `host` parameter here? Do we ever use it?

gaganawhad replied:
@benhutton I ended up using it here: https://github.com/desiringgod/desiring-god-new/blob/develop/lib/scripts/00083_update_facebook_count_on_non_canonical_wakes_locations.rb#L5. It has seemed helpful.

benhutton replied:
@gaganawhad shouldn't that have been the default host there, though?

gaganawhad replied:
@benhutton I couldn't use default host in order to test the query out locally and on staging, which have a different default host.

benhutton replied:
@gaganawhad should they? Does it make sense for wakes to have a different default host depending on environment?

gaganawhad replied:
No, it makes sense, but I wanted to run the script and validate it for Facebook metrics that show up only for production. For staging and local host they show up as nil or 0. Which is why I used an explicit host.

chrismaddocks replied:
@gaganawhad @benhutton If we prefer, there is some precedent for overriding the default host altogether. This might be a good practice in general when testing the social metrics (but not redirect) parts of Wakes.

https://github.com/desiringgod/wakes/blob/develop/spec/models/wakes/location_spec.rb#L36-L43

benhutton replied:
@gaganawhad @chrismaddocks something about all of these suggestions just feels wrong. EDD ftw.

1.) I think the naming collision between the `host` parameter of the `#url` method and the `#host` attribute of the `Wakes::Location` model is a code smell. What we probably want to do is call the parameter something like `override_host`. If we could do that, then the code here would be much easier to understand: `host = override_host || self.host || ENV['DEFAULT_HOST']`. I would prefer the explicitness we gain from this over an attempt to match things to the route helper API. This method should _almost never_ need to take any parameters.

2.) That then pushes us to look at the situations where we'd want to actually override the host. For redirects, you're going to be going off of paths unless a host is explicitly specified. You shouldn't need to override there. So it's only for metrics. And it seems like the use case for metrics is in scripts and console. In that case, you're probably going to want to override across the board, not just one by one. That is, you're not going to want a different host for every url, but the same host. So I wonder if there's some sort of global config that would be better suited, whether it's the `ENV` var or something that gets initially set to that var.

gaganawhad replied:
@benhutton good points! 

@chrismaddocks @benhutton 
I am an favor of using `host = override_host || self.host || ENV['DEFAULT_HOST']` as a next step, if not what Ben suggests in #2. I prefer calling it `host_override` instead of `override_host`.

benhutton replied:
that works for me. `host_override` sounds good.

benhutton commented on spec/wakes_spec.rb:
```diff
@@ -129,22 +176,39 @@ class WakeableModel < ActiveRecord::Base
         expect(Wakes::REDIS.get('/legacy-location-four')).to eq('/canonical-location-two')
       end
+
+      it 'doesn\'t store paths from other hosts in redis' do
```
I would either use double quotes so you don't have to escape the single quote OR expand the contraction to "does not"

chrismaddocks commented:
@benhutton - new build that addresses the PR feedback and tackles the redirects. I would like to write some integration tests to satisfy myself that it's all working as intended, but now would be a good review point.

benhutton commented on app/models/concerns/wakes/metrics/google_analytics_pageviews.rb:
```diff
@@ -18,9 +18,10 @@ def self.ordered_for_analytics_worker
-          where("(document->>'pageview_count_updated_through') IS NULL \
-                OR document->>'pageview_count_updated_through' < ?",
-                1.day.ago.to_date)
+          where('host IS NULL OR host = ?', [ENV['DEFAULT_HOST']]) # Only query GA for paths on the current host
```
@chrismaddocks need a spec for this

benhutton commented on lib/wakes/location_helper.rb:
```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+class LocationHelper
```
@chrismaddocks a question and a few thoughts here:

1.) Is this only used in the matcher?

2.) I'm generally not a fan of the name. 

3.) It looks like you are substantially re-implementing the `URI()` method. You might be able to get away with a clever use of `URI()` instead of this class, or you might look into making this class a thin wrapper around `URI()`

chrismaddocks replied:
@benhutton yes, only used in the matcher. It allows the spec to continue using strings to represent locations—preserving readability—but it extends it to also support domain names (as you probably already deduced).

For name, I had also considered `LocationString` but didn't love either name. Do you like that better? It's a lightweight wrapper, halfway between a string and a Wakes location. 

I'll take a look at wrapping `URI()`. I wanted to keep it lightweight since it's just for the spec, but my approach is definitely fragile.

benhutton replied:
play around with `URI()` and see what you think you can do with that. You may get by just putting a couple of helper methods in the matcher... Specifically

    def path(location_string)
      URI(location_string).path
    end
    
    def host(location_string)
      location_string = "http://#{location_string}" unless location_string.starts_with?('http')
      URI(location_string).host
    end

benhutton commented on spec/lib/wakes/redirect_mapper_spec.rb:
```diff
@@ -178,4 +178,66 @@
         .to have_wakes_graph(:canonical_location => '/target', :legacy_locations => ['/source'])
     end
   end
+
```
@chrismaddocks it looks like your code here is really just testing the `has_wakes_graph` matcher. Notice how all the other tests in this file have lines like `described_class.new '/source', '/target'`. That's what will actually test the redirect mapper.

benhutton commented on app/models/concerns/wakes/redirectors.rb:
```diff
@@ -16,7 +16,8 @@ def create_redirect_graph
         legacy_locations.reload
         legacy_paths_in_redis = legacy_locations.map do |legacy_location|
-          Wakes::REDIS.set(legacy_location.path, canonical_location.path) unless legacy_location.host.present?
+          canonical_path_or_url = canonical_location.host.blank? ? canonical_location.path : canonical_location.url
```
@chrismaddocks can we just call this variable `target`?

chrismaddocks replied:
@benhutton sure, I like that.

benhutton commented:
@chrismaddocks you have merge conflicts now. you need to rebase

And feel free to squash your last 3 commits in. They look good.

chrismaddocks commented:
@benhutton Ready for another look. Feedback has been incorporated. I also ran some tests from the rails console to create Wakes redirects to non-default domains, and it appears to behave as expected. I don't think I have more to do here, but would appreciate more Wakes-trained eyes!

cc: @joeosburn-desiringgod @gaganawhad

benhutton commented on app/models/wakes/location.rb:
```diff
@@ -3,17 +3,24 @@ class Wakes::Location < ActiveRecord::Base
 
   def label
-    path
+    (host + path if host.present?) ||
```
why not just use `path_or_url` here?

chrismaddocks replied:
@benhutton Could do. I was keeping it consistent with the abbreviated pseudo-URI syntax found elsewhere, namely `hostname/path`. This is not technically a URI (and the URI methods in Ruby do not parse it correctly). 

The shorter syntax helps with readability in [redirect_mapper_spec.rb](https://github.com/desiringgod/wakes/blob/01a02b9efa94278f5bbdc13da1615c7848b14ab2/spec/lib/wakes/redirect_mapper_spec.rb#L182-L232), and I thought the same considerations would apply when printing out the wakes graph. Do you still prefer the full URL?

benhutton replied:
@chrismaddocks what about something like `path_or_url.sub(%r{^https?://}, '')`? Basically, remove the protocol if it happens to be there.

benhutton commented on app/models/wakes/location.rb:
```diff
@@ -3,17 +3,24 @@ class Wakes::Location < ActiveRecord::Base
+  end
+
+  def path_or_url
+    (path if host.blank?) ||
```
if you want to be this concise, use a ternary operator: `host.blank? ? path : url`

I would probably flip it around to check presence, because that seems to read more logically to me: `host.present? ? url : path`

benhutton commented on app/models/wakes/resource.rb:
```diff
@@ -22,9 +22,10 @@ class Wakes::Resource < ActiveRecord::Base
   validates :label, :presence => true
 
   def to_s
+    legacy_location_labels = legacy_locations.pluck(:host, :path).map { |x| x.join('') }
```
can't either the new `label` or the new `path_or_url` method be used here?

chrismaddocks replied:
@benhutton I saw this pattern in the specs, and thought it was followed to avoid instantiating all of the legacy `Wakes::Location`s (since `pluck` retrieves _only_ `host` and `path` from the database) for debug code. It appears this happens anyway as part of a `Wakes.redirect`, though, so I'll revert back to the instance methods.

benhutton replied:
@chrismaddocks I like the instinct towards the performant code. In this particular case, though, because it's only for scripting and console usage, I would rather use DRY code than performant code.

benhutton commented on lib/wakes.rb:
```diff
@@ -39,8 +39,12 @@ def self.create!(*args)
+    RedirectMapper.new(source_path, target_path, label)
+  end
+
+  def self.redirect_with_hosts(source_host: nil, source_path:, target_host: nil, target_path:, label: nil)
```
@chrismaddocks I'm not a fan of this API.

Why not just use `self.redirect(source_path_or_url, target_path_or_url, label = nil)`?

Then add in logic in `RedirectMapper` to determine if a host is present or not.

A few reasons:

1.) It's a nicer API.
2.) It means that we don't have to maintain two different APIs.
3.) It means that devs don't have to remember two different APIs.

chrismaddocks replied:
@benhutton I was going to ask you about this on Friday if you were still around (this was definitely a decision point).

I went this way for now because I wasn't sure we wanted to extend this quasi-URI syntax all the way to the API. (Right now it is just confined to internal implementation and specs.) I'm not opposed to taking it the rest of the way, though.

benhutton replied:
@chrismaddocks what do you mean by "quasi-URI syntax"?

I wouldn't start from the syntax that you've been using elsewhere. I would start from the question of, "what's the ideal API for this?" and then figure out how to implement it. You probably should support a few different cases:

    Wakes.redirect '/source', '/target'
    Wakes.redirect 'www.domain.com/source', '/target'
    Wakes.redirect 'http://www.domain.com/source', '/target'

chrismaddocks replied:
@benhutton `www.domain.com/source` would be an example. It's not a valid URI, and the Ruby URI method chokes on it, hence "quasi-URI".

I like it for readability in the specs, but wasn't sure I was ready to make it an API.

Anyway, with the code in its present state, we now support all three cases (as well as the reverse directions).

benhutton replied:
gotcha. So from an API standpoint, we don't care what Ruby's `URI` thinks about it :). We can put it in the API... it'll be more useful than the one that includes the protocol. We're going to strip out the protocol, anyways.

benhutton commented on lib/wakes/matchers.rb:
```diff
@@ -1,20 +1,42 @@
+  def build_location_string(location)
+    if location.present
+      location.host + '/' + location.path if location.host.present?
+      location.path
```
line 18 here will always get returned. Thus, line 17 is useless.

benhutton commented on lib/wakes/matchers.rb:
```diff
@@ -1,20 +1,42 @@
+  end
+
+  def build_location_string(location)
+    if location.present
```
this should be `location.present?`, with the `?`

benhutton commented on lib/wakes/matchers.rb:
```diff
@@ -1,20 +1,42 @@
+    URI(location_string).path
+  end
+
+  def build_location_string(location)
```
why not use either `location#path_or_url` or `location#label`?

benhutton commented on lib/wakes/matchers.rb:
```diff
@@ -1,20 +1,42 @@
+      wakes_resource.canonical_location.host == host(canonical_location) &&
+      wakes_resource.canonical_location.path == path(canonical_location) &&
+      wakes_resource.legacy_locations.pluck(:host, :path).sort ==
+        legacy_locations.map { |x| [host(x), path(x)] }.sort
```
you use this twice, so might as well make it a method

benhutton commented on lib/wakes/matchers.rb:
```diff
@@ -1,20 +1,42 @@
+
 RSpec::Matchers.define :have_wakes_graph do |canonical_location:, legacy_locations: []|
+  def host(location_string)
+    location_string = "http://#{location_string}" if URI(location_string).scheme.nil?
```
you use this twice, so might as well make it a method

benhutton replied:
specifically, 

    def uri(location_string)
      location_string = "http://#{location_string}" if URI(location_string).scheme.nil?
      URI(location_string)
    end

Then `#host` and `#path` just need to call `uri(location_string).path` and `uri(location_string).host`

benhutton commented on lib/wakes/redirect_mapper.rb:
```diff
@@ -1,28 +1,28 @@
-        puts resource.to_s
-      end
+      Wakes::Resource.where(:id => resources_affected).each { |resource| puts resource.to_s }
+      puts '' if options[:newline]
```
this seems a little over-engineered

chrismaddocks replied:
@benhutton I meant to discuss this with you. Rubocop was rejecting this class because it is too long, but I did not see an obvious place to trim it. This is admittedly a hack but it seemed this would be true anywhere I trimmed. 

My preference was actually to use parallel assignment [here](https://github.com/desiringgod/wakes/blob/01a02b9efa94278f5bbdc13da1615c7848b14ab2/lib/wakes/redirect_mapper.rb#L7-L10), but I see that is discouraged in the style guide and Rubocop. Alternatively, we could drop the [bold](https://github.com/desiringgod/wakes/blob/01a02b9efa94278f5bbdc13da1615c7848b14ab2/lib/wakes/redirect_mapper.rb#L14).

Open to suggestions here. I'm dancing for Rubocop but don't actually think the class needs refactoring.

benhutton replied:
yeah, at some point I need to just refactor this class down into a bunch of smaller classes. Feel free to add in the rubocop comments to ignore class length cop.

benhutton commented on spec/lib/wakes/redirect_mapper_spec.rb:
```diff
@@ -174,8 +174,61 @@
+        expect(Wakes::Location.find_by(:host => 'some.host', :path => '/target').resource)
+          .not_to have_wakes_graph(:canonical_location => '/target', :legacy_locations => ['/source'])
+        expect(Wakes::Location.find_by(:host => nil, :path => '/source').resource)
+          .to have_wakes_graph(:canonical_location => 'some.host/target', :legacy_locations => ['/source'])
```
aren't all three of these expectations working on the same resource? If you really wanted, you could do `expect(Wakes::Location.find_by(:host => nil, :path => '/source').resource_id).to eq(Wakes::Location.find_by(:host => 'some.host', :path => '/target').resource_id)

Once you've proven that, it's unnecessary to have both the `to` and the `not_to` expectations here. If it has one wakes graph, it necessarily doesn't have the other.

chrismaddocks replied:
@benhutton Fixed. I had read this as testing membership instead of equivalence.

benhutton replied:
@chrismaddocks this looks the same to me

chrismaddocks replied:
@benhutton You're right, I missed one, sorry. Will be fixed in next push

benhutton commented on spec/lib/wakes/redirect_mapper_spec.rb:
```diff
@@ -174,8 +174,61 @@
+
+    context 'source is default, target is non-default' do
+      it 'points the source location to the target resource' do
+        described_class.new '/source', '/target', 'New Label', nil, 'some.host'
```
following the other comment about API, this would be a lot nicer if it was `describe_class.new '/source', 'some.host/target', :label => 'New Label'`

benhutton commented on lib/wakes/redirect_mapper.rb:
```diff
@@ -33,11 +35,11 @@ def resources_affected
 
     def target_location
-      @target_location ||= Wakes::Location.find_by(:path => target)
+      @target_location ||= Wakes::Location.find_by(:host => host(target_path), :path => path(target_path))
```
@chrismaddocks wouldn't it be simpler to just setup `target_host` and `source_host` variables in the `#initialize` method? ALSO, doesn't `path(target_path)` seem wrong to you? If it says `target_path`, it should be the path!

You could also do something like the following:

    def initialize(source_path_or_url, target_path_or_url, label = nil)
      @source_path_or_url = source_path_or_url
      @target_path_or_url = target_path_or_url
      ...
    end

    def source_path
      path(@source_path_or_url)
    end

    def source_host
      host(@source_path_or_url)
    end

    def target_path
      ...

chrismaddocks replied:
@benhutton You may recall that I had @target_host and @source_host variables in the last check-in. Because we are now running into Rubocop class length restrictions on `RedirectMapper`, I decided to condense these into two variables, since the other two didn't add much to readability. 

I also had @source_path_or_url before pushing the commits, but ran into line length issues (which couldn't be resolved without introducing class length issues).

Much of the stuff you're pointing out today is monkeying I did consciously to fit within the style guide. Let me relax the class length cop and then I think we'll be on the same page.

benhutton commented on lib/wakes/location_string.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module LocationString
```
@chrismaddocks if you're going to go this route (which makes sense, since you need it in the `RedirectMapper` code too, I would probably rather go with a class. The module name here is pretty bad, and I think the difficulty ended up with a good name is because it needs to be a class.

I would probably do the following

    class URIFromLocationString
      def initialize(location_string)
        location_string = "http://#{location_string}" if URI(location_string).scheme.nil?
        URI(location_string)
      end
    end

Then you use it like `URIFromLocationString.new('/some/path').host`

chrismaddocks replied:
@benhutton Won't `URIFromLocationString.new` return an instance of URIFromLocationString, rather than a URI? 

I think it would need to be
```
class URIFromLocationString
  def self.generate(location_string)
    location_string = "http://#{location_string}" if URI(location_string).scheme.nil?
    URI(location_string)
  end
end
```

and then `URIFromLocationString.generate('/some/path').host`

benhutton replied:
@chrismaddocks yes! you are correct

chrismaddocks replied:
@benhutton What do you think about this syntax for `RedirectMapper.initialize`

```
      @source_host, @source_path = URIFromLocationString.get_host_and_path(source_path_or_url)
      @target_host, @target_path = URIFromLocationString.get_host_and_path(target_path_or_url)
```

and this

```
class URIFromLocationString
  def self.generate(location_string)
    location_string = "http://#{location_string}" if URI(location_string).scheme.nil?
    URI(location_string)
  end

  def self.get_host_and_path(location_string)
    uri = self.generate(location_string)
    uri.host, uri.path
  end
end
```

benhutton replied:
@chrismaddocks I like that!

chrismaddocks commented:
@benhutton Ignore the "revert to non-Rubocop..." commit, it is reverted in the later commits and will be squashed. This ended up being unnecessary after refactoring `LocationString`.

benhutton commented:
@chrismaddocks this is all looking amazing to me. Thanks for hanging with me through so many tweaks. This is some very solid code here. Feel free to start squashing down as soon as you're ready.

chrismaddocks commented:
@benhutton Excellent! Thanks for your thorough help in polishing it.

benhutton commented:
@chrismaddocks what's the status here? Are you done? I'd like to get a lot of these commits squashed before we merge.

chrismaddocks commented:
@benhutton will be squashing this morning!
On Thu, Apr 28, 2016 at 6:20 AM Ben Hutton <notifications@github.com> wrote:

> @chrismaddocks <https://github.com/chrismaddocks> what's the status here?
> Are you done? I'd like to get a lot of these commits squashed before we
> merge.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/desiringgod/wakes/pull/22#issuecomment-215420390>
>

chrismaddocks commented:
@benhutton OK, this should be ready to go now!

benhutton commented:
@chrismaddocks this is merged. You can now submit a PR that updates the app to the latest version of the gem (which can of course be part of a bigger PR that you're working on)
